### ISSUE TITLE
JAR constructor filter is not being respected for certain classes

### ISSUE THREAD

**Issue by yuniorb** [lukehutch/fast-classpath-scanner#100]
I am seeing an unexpected behavior with the jar filter and how the scanner loads the classes within jar files. Attached is a zip folder with three projects `dep-c`, `dep-w`, and `scan`. The first two projects generate a jar file that I manually copy to the scan project where I am using fast-class-scanner to get the type of a method parameter found in one of the jars. The unexpected behavior is that even when a jar filter is provided to the scanner (e.g `jar:dep-w.jar`), the wrong class is loaded by the MethodAnnotationMatchProcessor. This only seems to happen when both classes have the same name and are under the same namespace.

To replicate the behavior run the test `testJarFilter` under the scan project. 

Here is the verbose output from FCS: [scanner-log.txt](https://github.com/lukehutch/fast-classpath-scanner/files/706372/scanner-log.txt)

[Repro steps projects](https://github.com/yuniorb/FCS-Issue)

**Comment by lukehutch** [comment]
Hi @yuniorb,

Thanks for the bug report -- this was by far the most complete and well-specified bug report I have received so far, thanks for the log and the minimal testcase!

You uncovered a really interesting bug. What is going on here is that the `MatchProcessor` was being called with the correct class name (from `dep-w.jar`), however `dep-c.jar` was earlier on the classpath than `dep-w.jar`, so right before the `MatchProcessor` was passed the class reference, the `ClassLoader` handling both `dep-c.jar` and `dep-w.jar` ended up loading the `-c` version of the class, rather than the `-w` version.

This is pretty tricky to fix correctly. What I did was create a new `URLClassLoader` for each classpath element (dir or jar). Each of these new `URLClassLoaders` inherits the parent (JRE) `ClassLoader`, so they should be able to resolve `java.lang.*` etc., but they contain only one URL, for a single classpath element. This means that for standalone jars, the patch I just checked in should work fine, but if a class contains references to non-system classes defined outside its own jar, you'll get a `LinkageError`, since the other jars on the classpath are not visible. (I caught `LinkageError`, and in this situation, FCS falls back to trying to load the class by calling all `ClassLoaders` in order.)

The reason this is difficult to fix in the general case (or actually, potentially impossible in the general case) is that not all `ClassLoaders` are `URLClassLoaders` -- otherwise I could just include all classpath URLs in a single unified `URLClassLoader`, and change the order as needed to ensure that each `MatchProcessor` resolves its own matched class first.

(To summarize -- this is not a complete fix, but to trigger this bug you would have to have multiple classes of the same name on the classpath, and those classes would have to refer to non-JRE classes outside their own jarfile.)

Anyway, the fix I just pushed may be good enough to resolve the problem for your usecase -- please test and let me know.

The only *correct* fix for this, for your usecase, is to override the classpath to include only `dep-w.jar`, and not `dep-c.jar`.

**Comment by lukehutch** [comment]
Sorry, closed this accidentally. Please let me know if this is fixed for you.

**Comment by yuniorb** [comment]
Thanks for the prompt reply and bug fix. I'll try it out the fix this afternoon at the latest.

**Comment by yuniorb** [comment]
Awesome! This fix the issue.

**Comment by yuniorb** [comment]
Hi @lukehutch 

This fix has some side-effects that might be a little problematic for some use cases. I have a maven project. When this project is packaged it will have two jars `project.jar` and `project-tests.jar`. I am  loading both jars into a class loader and adding them to FCS, filtering down to to just the test jar using a `scanSpec` for the scan.  Any `ClassAnnotationMatchProcessor` that gets hit will provide for you the class/method that matches. Before the fix you could call `matchingClass.getAnnotation(Foo.class);` without an issue because it would find it in the classpath, but now because it is scoped to just the jar this would throw the following exception:

`java.lang.ArrayStoreException: sun.reflect.annotation.TypeNotPresentExceptionProxy` [exception details](https://github.com/lukehutch/fast-classpath-scanner/files/706902/TypeNotPresentExceptionProxy.txt)

I could probably get around the original issue by using a class loader to only load `dep-w.jar`, `dep-w-tests.jar` complete the scan and then create another FCS instance with a different class loader containing `dep-c.jar`, `dep-c.-tests.jar` and so on for all the jars. 

I don't see a way to get around the fix for the use case I mentioned thouhgt, any ideas?

**Comment by lukehutch** [comment]
Hi @yuniorb,

I committed two changes in this issue. The first change had exactly the problem you just described -- I created a new `ClassLoader` for every jar that contained only the one jar's URL. The second change however created a new `ClassLoader` for each jar, consisting of all URLs from all jars, but with the URL of the given jar moved to the beginning of the list of URLs for just that `ClassLoader`. In other words, given URLs `file:/path/a.jar`, `file:/path/b.jar` and `file:/path/c.jar`, three ClassLoaders would be created, but the first would resolve in the order a, b, c; the second would resolve in the order b, a, c; and the third would resolve in the order c, a, b.

So the second change (the current master version) should in fact be able to find the annotation `Foo.class` in your example. Can you please call the following inside your `MatchProcessor` to see what URLs are visible? (This is off the top of my head, you may have to tweak it a bit if it's not exactly correct..)

```java
System.out.println(Arrays.toString(
    ((URLClassLoader) matchingClass.getClassLoader()).getURLs()
));
```

**Comment by yuniorb** [comment]
Here is the output

`[file:/Users/yuniorb/projects/jars/project-tests.jar, file:/Users/yuniorb/projects/project/target/classes/]`

This is how I am loading the jars into FCS: 
`  // Add all the jars under the provided path
    File folder = new File(args[0]);

    for (File sub : folder.listFiles()) {
      if (sub.isFile() && FilenameUtils.getExtension(sub.getName()).equalsIgnoreCase("jar")) {
        urls.add(sub.toURI().toURL());
      }

     URLClassLoader loader = new URLClassLoader(urls.toArray(new URL[urls.size()]));
     // then add loader to FCS
    `

**Comment by lukehutch** [comment]
When you say "then add loader to FCS", do you mean by calling the following?:

```java
new FastClasspathScanner().overrideClassLoaders(loader).scan();
```

**Comment by lukehutch** [comment]
Also, since the output from the test I requested showed that `project-tests.jar` is in the classpath, can you please double-check that the annotation you're trying to get is actually in that package?

**Comment by yuniorb** [comment]
I updated the example to clarify, but I am not using the `overrideClassLoaders` I am using the `addClassLoader`. The annotation is in the the `project-tests.jar`.

**Comment by lukehutch** [comment]
Hmm -- that's really strange that `matchingClass`'s classloader couldn't find the annotation then, since it's just a standard `URLClassLoader`, and it does in fact have a reference to `project-tests.jar` internally.

Can you try the following please (in main, not in the `MatchProcessor`; again, this is untested code, you might have to fix something):

```java
URLClassLoader loader = new URLClassLoader(new URL[] {
    new URL("file:/Users/yuniorb/projects/jars/project-tests.jar",
    new URL("file:/Users/yuniorb/projects/project/target/classes/")
});
Class<?> annotation = Class.forName("pkg.Foo", true, loader);
System.out.println(annotation);
```

**Comment by yuniorb** [comment]
The code above work both in main and in the match processor

**Comment by lukehutch** [comment]
OK, can you please take a close look at the URLClassLoader in your MatchProcessor to see if it looks different from the example above? (Maybe take a look at their internal fields in a debugger -- they are initialized in the same way, with those two same  URLs!)

**Comment by yuniorb** [comment]
I'll take a look

**Comment by lukehutch** [comment]
Thanks. Also please try re-creating the same situation directly, outside of the `MatchProcessor:`

```java
URLClassLoader loader = new URLClassLoader(new URL[] {
    new URL("file:/Users/yuniorb/projects/jars/project-tests.jar"),
    new URL("file:/Users/yuniorb/projects/project/target/classes/")
});
// Replace "MatchingClass" here with a class matched by the MatchProcessor
Class<?> matchingClass = Class.forName(MatchingClass.class.getName(), true, loader);
System.out.println(matchingClass.getAnnotation(Foo.class));
```

If that works, please try changing `true` to `false` in the code above.

If both of those work, please try again (with both `true` and `false`) from inside the `MatchProcessor`.

If that works, please let me know if the debugger results for the two different URLClassLoaders.

If that doesn't reveal any differences, I would really appreciate it if you could please create the most minimal testcase you can think of that can demonstrate the problem with the latest git master version of FCS.

Thanks for helping me this debugged!

**Comment by yuniorb** [comment]
[Here](https://github.com/lukehutch/fast-classpath-scanner/files/707097/loaderDiff.txt) is the difference between the loaders

**Comment by lukehutch** [comment]
They look basically identical on the surface, although are the `packages` HashMaps basically the same between the two of them? (I don't know enough about the internal structure of `URLClassLoader` to suggest anything other than that.)

The only thing I can think of is that boolean parameter I suggested changing from `true` to `false` in my previous comment -- that specifies whether or not a class is initialized when loaded. It is set to `false` when loading a class for a `MatchProcessor`. However, the initializer is supposed to be called the first time the class is used, so it shouldn't affect linking. Other than that, as far as I know, these two `ClassLoaders` should be *identical*. I'm kind of stumped at this point. If going through that previous comment doesn't turn up anything else, I'll try debugging a minimal testcase if you're able to provide one.

**Comment by yuniorb** [comment]
The interesting thing is that I have three match processors. Two of them work fine. All the failures only happened with a specific annotation. Once I changed the class loader from the one being provided by the matched class to the original URLClassLoader it all worked fine. Let me try changing the parameter to see if it makes any difference.

**Comment by yuniorb** [comment]
Setting the boolean parameter to true or false doesn't make a difference inside the match processor.

**Comment by yuniorb** [comment]
Should the urls from both loaders be the same or does the `MatchProcessor` matched class loader should have less?

**Comment by lukehutch** [comment]
The result you get from `URLClassLoader#getURLs()` should match the classpath that you expect. The example code I gave you just took the output of that function that you gave me earlier, and created a new `URLClassLoader` from exactly the same URLs. And suddenly it works! That is the bizarre thing about this. I honestly have no idea what could be causing this problem.

**Comment by lukehutch** [comment]
Is it possible that that same annotation is defined in multiple places with exactly the same name? What may be happening is that earlier the system classloaders have *already located* that annotation, and resolved them to a specific classpath element (jar or directory) that is *not* one of the two that the `URLClassLoader` knows about. The copy of `project-tests.jar` that the `URLClassLoader` knows about contains a *second copy* of the annotation class, and the `project/target/classes/` directory does not contain a copy of it. For some reason, the `MatchProcessor`'s `ClassLoader` is trying to pull the class from an external URL, which is not in the `ClassLoader`, and the loading fails.

**Comment by lukehutch** [comment]
Hi @yuniorb,

OK, I looked into this extensively, and figured out what is happening here.

The problem is that the JRE system classloaders actually cache classes after loading them. So if your classpath contains `a.jar` and `b.jar` in that order, and both jars contain a definition of a class `X`, then if somehow you end up loading class `X` before scanning, there is no way to cause the system classloaders to "unload" class `X`, and even if you create a new `URLClassLoader` that points to `b.jar` only, when you try loading `X` from `b.jar`, you are passed the cached instance in the parent classloader.

It is possible to dynamically load and reload classes, by creating a custom dynamic classloader that reads the bytes in `X.class` from `b.jar`, and then instantiates a class from those bytes. However, if FastClasspathScanner does this, and passes a `Class<?>` reference to a `MatchProcessor`, then the `MatchProcessors` will be dealing with a *different class* than the main program that invoked `FastClasspathScanner`. If users of FCS pass class references back and forth between the main program and a `MatchProcessor`, or if the two different class definitions differ, some very weird and hard-to-trace bugs could result.

Consequently, I made the decision to restrict FCS such that the behavior of what it matches is always consistent with the `ClassLoader`(s) of its caller. What that means for this example is that including `"jar:b.jar"` in the scan spec will mean that the class `X` will not be found, *because that class was masked by an earlier definition of a class of the same name, in `a.jar`, which is not whitelisted*.

This may not be the behavior you want, but it is the only way to avoid surprises here. The "fix" for this, in your case, is to set the classpath before starting the JRE so that classes with a given name are defined in only one jar or directory.

Note that even using `FastClasspathScanner#overrideClasspath(...)` or `FastClasspathScanner#overrideClassLoaders(...)` suffer from this same issue of class caching, and could lead to unexpected results if you somehow accidentally load the wrong instance of a given multiply-defined class before starting the scan.

FCS now sets up a scan in the following order:

1. All `ClassLoaders` are queried for a list of URLs they scan. This produces an ordered list of classpath elements (jarfiles and directories).
2. The recursive directory listing of files within each classpath element is obtained.
3. The paths of files within each classpath element are filtered against class and package whitelists and blacklists.
3. "Classpath masking" is used to hide the second and subsequent references to classfiles or other files in the classpath that have matching paths.
4. Scan spec criteria for jar whitelisting and blacklisting are applied, as well as `"-jar:"` and `"-dir:"`, which disables jar and directory scanning altogether.
5. With any remaining classpath elements, `MatchProcessors` are called for each matched classfile or filename.

In other words, step 4 is now performed after step 3, whereas before it was the other way around.

Sorry if this is not the clean solution you were looking for, but introducing dynamic classloading would be the only solution, and would create a huge amount of additional complexity and sources of potential bugs for users.

**Comment by lukehutch** [comment]
One other note about this -- I detailed [how to handle dynamic classloading](https://github.com/lukehutch/fast-classpath-scanner/wiki/4.-Working-with-nonstandard-ClassLoaders#dynamic-or-manual-classloading) in the documentation, but these same two solutions also apply to your situation, where you want to control specifically which definitions of a given named class are loaded.

**Comment by lukehutch** [comment]
I'm going to have to revert part of this bugfix due to an impact on scanning performance, as described in #117. Classpath masking must be performed after applying jar whitelist/blacklist criteria, not before, otherwise the directory entries for every jar in the classpath must be read (including those of non-whitelisted and blacklisted jars) before classpath masking is applied (which can increase scan time in a non-insignificant way for large classpaths).

What this means is that if you use a whitelist, FastClasspathScanner will give you info about any matching class in the whitelist, but all bets are off as to whether this info will actually match a class loaded by a classloader, since the classloader can end up loading a class from outside the whitelist (e.g. if the whitelisted jar occurs later in the classpath than another jar that defines the same class).

In summary, using jar whitelisting/blacklisting when there are multiple definitions of the same class will have undefined behavior.
